### PR TITLE
#81 - edit middleware for auth users - Aisha Athman Lali

### LINKED ISSUE #81: Update auth middleware following rules in description
`admin` - can access any endpoint
`POST/sign-in` - can be accessed by anybody
`POST /tender` - Can be accessed by admins and buyers
`GET - /skills` - can be accessed by anybody with a valid auth token
`GET /buyer-tender` - can be accessed by anybody with a valid auth token
`GET /bidder-bid` - can be accessed by anybody with a valid auth token
`GET /tenders` - can be accessed by anybody with a valid auth token

### PR DESCRIPTION
**Pull request by aishaathmanlali**
## Naming Rules

Name your PR like this: ISSUENUMBER-TITLE-YOURNAME

## Description

Write a summary of your change. Explain why you have made it.

## Related to

Make sure you include the issue number with a hash sign # so Github can link this PR to the right issue, like this:

Fixes #1

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have carefully reviewed my own code
- [ ] I have commented my code
- [ ] I have updated any documentation

### CODE REVIEW

File: server/api.js [hunk 1]
```diff
@@ -51,7 +47,30 @@ const auth = async (req, res, next) => {
...
+				"/skills",
+				"/buyer-tender",
+				"/bidder-bid",
+				"/tenders",
+			];
+			if (publicPaths.includes(path)) {
+				return next();
+			}
+		}
+		return res.status(403).json({ code: "FORBIDDEN" });
```
Comment (soonick) [inline server/api.js]: This is basically what we want to do, but having a map that allows us to add routes without having to add more code is beneficial for the future.

Do something similar to what I suggested here for the map: https://github.com/LauraSantiag0/Love-Me-Tender/issues/77

And combine it with the logic you have here

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js (hunk 1 above)
Comment (aishaathmanlali) [reply server/api.js]: working on it

File: server/api.js [hunk 2]
```diff
@@ -7,51 +7,78 @@ const itemsPerPage = 25;
...
-		if (allowlist[method] && allowlist[method].includes(path)) {
+		const allowedAccess = allowlist[method] && allowlist[method][path];
+
+		if (allowedAccess === "public") {
 			return next();
 		}
 
-		const authHeader = req.headers.authorization;
-		const token = authHeader?.split(" ")[1];
+		if (allowedAccess === "token") {
```
Comment (soonick) [inline server/api.js]: Admins should have access to everything even if it's not listed in the allowlist

File: server/api.js [hunk 3]
```diff
@@ -7,51 +7,78 @@ const itemsPerPage = 25;
 const router = Router();
 
 const allowlist = {
-	POST: ["/sign-in"],
+	POST: {
+		"/sign-in": "public",
+		"/tender": "token",
```
Comment (soonick) [inline server/api.js]:
```suggestion
		"/tender": "buyer",
```

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js [hunk 4]
```diff
@@ -51,7 +62,19 @@ const auth = async (req, res, next) => {
...
+		if (user.role === "admin") {
+			return next();
+		}
+
+		if (method === "POST" && path === "/tender") {
+			if (user.role === "buyer") {
+				return next();
+			}
+			return res.status(403).json({ code: "FORBIDDEN" });
+		}
```
Comment (soonick) [inline server/api.js]:
```suggestion
		if (allowedAccess === user.role) {
			return next();
		}
```

**Review by soonick — CHANGES_REQUESTED** [review]

**Review by soonick — APPROVED** [review]

**Pull request merged**